nat-lab/bin/netcat.py

```diff
@@ -0,0 +1,209 @@
+
+    def _register_stdin(self):
+        """Register stdin to the selector for read events, unless disabled"""
+        if not self.args.d:
+            self.selector.register(
+                sys.stdin, selectors.EVENT_READ, self._read_from_stdin
+            )
+
+    def _read_from_socket(self):
+        """Handle incoming data from the socket"""
+        data, addr = self.sock.recvfrom(RECV_SIZE)
+        if self.udp and self.listen and not self.client_addr:
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.client_addr = addr
+            self.sock.connect(addr)
+        if data:
+            sys.stdout.buffer.write(data)
+            sys.stdout.flush()
+        else:
+            # Connection closed
+            self.selector.unregister(self.sock)
+            self.should_close = True
+
+    def _read_from_stdin(self):
+        """Handle input from stdin and send it over the socket"""
+        data = sys.stdin.buffer.readline()
+        if data:
+            if self.listen and self.client_addr:
+                self.sock.sendto(data, self.client_addr)
+            else:
+                self.sock.send(data)
+        else:
+            # EOF on stdin
+            self.selector.unregister(sys.stdin)
+            self.should_close = True
+
+    def run(self):
+        try:
+            if self.args.l:
+                self._listen()
+            else:
+                self._connect()
+            if not self.args.z:
+                self._register_socket()
+                self._register_stdin()
+                while not self.should_close:
+                    events = self.selector.select(timeout=None)
+                    for key, _ in events:
+                        callback = key.data
+                        callback()
+        except KeyboardInterrupt:
+            print("")
+            sys.exit(2)
+        except OSError as e:
+            if e.errno in (errno.EBADF, errno.ECONNREFUSED):
+                # connection closed
+                sys.exit(0)
+            print(f"nc: {e}", file=sys.stderr)
+            sys.exit(1)
+        finally:
+            self.selector.close()
+            if self.sock:
+                self.sock.close()
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Netcat clone in Python")
+    parser.add_argument("-6", dest="ipv6", action="store_true", help="Use IPv6")
+    parser.add_argument(
+        "-4", dest="ipv4", action="store_true", help="Use IPv4 addresses"
+    )
+    parser.add_argument("-v", action="store_true", help="Verbose mode")
+    parser.add_argument("-n", action="store_true", help="Do not resolve hostnames")
+    parser.add_argument("-l", action="store_true", help="Listen mode (server)")
+    parser.add_argument("-u", action="store_true", help="UDP mode")
+    parser.add_argument("-d", action="store_true", help="Do not read from stdin")
+    parser.add_argument(
```

Comment:
IMHO we can remove `-z` as it is not supported, and passing it, causes silent closure.

Reply:
I think that is the expected behavior, no? -z will exit "successfully" only if there is an open port and it succeeds connecting, otherwise it fails with connection refused.

Reply:
sorry, I got mixed up with the meaning of `-z`